lib/saml_idp/request.rb

```diff
@@ -31,48 +31,129 @@ def initialize(raw_xml = "")
       self.raw_xml = raw_xml
     end
 
+    def logout_request?
+      logout_request.nil? ? false : true
+    end
+
+    def authn_request?
+      authn_request.nil? ? false : true
+    end
+
     def request_id
-      authn_request["ID"]
+      if authn_request?
+        authn_request["ID"]
+      elsif logout_request?
+        logout_request["ID"]
+      end
+    end
+
+    def requested_authn_context
+      if authn_request? && authn_context_node
+        authn_context_node.content
+      else
+        nil
+      end
     end
 
     def acs_url
       service_provider.acs_url ||
         authn_request["AssertionConsumerServiceURL"].to_s
     end
 
+    def logout_url
+      service_provider.assertion_consumer_logout_service_url
+    end
+
+    def response_url
+      if authn_request?
+        acs_url
+      elsif logout_request?
+        logout_url
+      end
+    end
+
+    def logger(msg)
+      if Rails && Rails.logger
+        Rails.logger.info msg
+      else
+        puts msg
+      end
+    end
+
     def valid?
-      service_provider? &&
-        valid_signature? &&
-        acs_url.present?
+      # TODO: This should validate against the schema.
+
+      unless service_provider?
+        logger "Unable to find service provider for issuer #{issuer}"
+        return false
+      end
+
+      unless (authn_request? ^ logout_request?)
+        logger "One and only one of authnrequest and logout request is required. authnrequest: #{authn_request?} logout_request: #{logout_request?} "
+        return false
+      end
+
+      unless valid_signature?
+        logger "Signature is invalid in #{raw_xml}"
+        return false
+      end
+
+      if response_url.nil?
+        logger "Unable to find response url for #{issuer}: #{raw_xml}"
+        return false
+      end
+
+      return true
     end
 
     def valid_signature?
-      service_provider.valid_signature? document
+      # Force signatures for logout requests because there is no other
+      # protection against a cross-site DoS.
+      service_provider.valid_signature?(document, logout_request?)
     end
 
     def service_provider?
       service_provider.valid?
     end
 
     def service_provider
-      @service_provider ||= ServiceProvider.new((service_provider_finder[issuer] || {}).merge(identifier: issuer))
+      @_service_provider ||= ServiceProvider.new((service_provider_finder[issuer] || {}).merge(identifier: issuer))
     end
 
     def issuer
-      @content ||= xpath("//saml:Issuer", saml: assertion).first.try(:content)
-      @content if @content.present?
+      @_issuer ||= xpath("//saml:Issuer", saml: assertion).first.try(:content)
```

Comment:
I believe there was a reason for the `@content if @content.present?` maybe append a `presence` to the call so we have no opportunity to output a blank string or something.